File: apps/mobile/android/app/src/main/java/cc/onedesk/mobile/MobileGatewayClient.kt
package cc.onedesk.mobile

import android.os.Build
import org.json.JSONArray
import org.json.JSONObject
import java.io.Closeable
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

class MobileGatewayClient(
    private val deviceId: () -> String,
    private val stableDeviceKey: () -> String,
    private val logs: MobileLogStore,
    private val onSchemeEvent: (desktop: JSONObject, descriptor: JSONObject, eventId: String) -> Boolean,
    private val onJsApiEvent: (capability: String, payload: JSONObject, requestId: String, sourceKey: String) -> JSONObject,
) : Closeable {
    private val transportLock = Any()
    private val eventExecutor = Executors.newSingleThreadExecutor { task ->
        Thread(task, "OneDesk-QuicEvents").apply { isDaemon = true }
    }
    private val scheduler = Executors.newSingleThreadScheduledExecutor { task ->
        Thread(task, "OneDesk-QuicHeartbeat").apply { isDaemon = true }
    }
    private val stateMachine = MobileConnectionStateMachine()
    private val running = AtomicBoolean(false)
    @Volatile private var transport: MsQuicNativeTransport? = null
    @Volatile private var endpointKey: String? = null
    @Volatile private var activeDesktop: JSONObject? = null
    @Volatile private var heartbeat: ScheduledFuture<*>? = null
    @Volatile private var reconnectFuture: ScheduledFuture<*>? = null
    @Volatile private var reconnectAttempt = 0

    fun request(
        host: String,
        port: Int,
        request: JSONObject,
        timeoutMs: Int = 7_000,
        expectedFingerprint: String? = null,
    ): JSONObject {
        val requestId = request.optString("requestId").ifBlank { "req-${UUID.randomUUID()}" }
        request.put("requestId", requestId)
        val client = ensureTransport(host, port, expectedFingerprint, timeoutMs)
        val envelope = JSONObject()
            .put("protocolVersion", OneDeskProtocol.PROTOCOL_VERSION)
            .put("messageType", GatewayMessageType.REQUEST.wireValue)
            .put("messageId", requestId)
            .put("correlationId", JSONObject.NULL)
            .put("payload", request)
        val responseBytes = client.request(envelope.toString().toByteArray(Charsets.UTF_8), timeoutMs)
        val responseEnvelope = JSONObject(String(responseBytes, Charsets.UTF_8))
        if (responseEnvelope.optString("messageType") != GatewayMessageType.RESPONSE.wireValue ||
            responseEnvelope.optString("correlationId") != requestId
        ) {
            throw IllegalStateException("GatewayResponseCorrelationMismatch")
        }
        return responseEnvelope.getJSONObject("payload")
    }

    fun serverFingerprint(): String = synchronized(transportLock) {
        transport?.serverFingerprint() ?: throw IllegalStateException("GatewaySessionOffline")
    }

    fun uploadLog(desktop: JSONObject, entry: JSONObject): Boolean {
        val trustCredential = desktop.optString("trustCredential")
        if (trustCredential.isBlank()) return false
        return runCatching {
            request(
                desktop.getString("host"),
                desktop.optInt("port", 48320),
                authorizedRequest("logs", trustCredential).put("logs", JSONArray().put(entry)),
                timeoutMs = 2_500,
                expectedFingerprint = desktop.optString("gatewayFingerprint"),
            ).optBoolean("ok")
        }.getOrDefault(false)
    }

    fun startSubscription(desktop: JSONObject) {
        val trustCredential = desktop.optString("trustCredential")
        if (trustCredential.isBlank()) return
        running.set(true)
        activeDesktop = JSONObject(desktop.toString())
        heartbeat?.cancel(true)
        val response = request(
            desktop.getString("host"),
            desktop.optInt("port", 48320),
            authorizedRequest("subscribe", trustCredential),
            expectedFingerprint = desktop.optString("gatewayFingerprint"),
        )
        if (!response.optBoolean("ok")) {
            running.set(false)
            throw IllegalStateException(response.optString("message", "无法订阅方案更新"))
        }
        reconnectAttempt = 0
        reconnectFuture?.cancel(false)
        reconnectFuture = null
        endpointKey?.let { key ->
            if (stateMachine.state.phase == MobileConnectionPhase.Synchronizing) {
                stateMachine.synchronized(key)
            }
        }
        heartbeat = scheduler.scheduleWithFixedDelay(
            { sendHeartbeat() },
            15,
            15,
            TimeUnit.SECONDS,
        )
    }

    fun stopSubscription() {
        running.set(false)
        heartbeat?.cancel(true)
        heartbeat = null
        reconnectFuture?.cancel(true)
        reconnectFuture = null
        reconnectAttempt = 0
        activeDesktop = null
        stateMachine.disconnect()
        synchronized(transportLock) {
            transport?.close()
            transport = null
            endpointKey = null
        }
    }

    fun authorizedRequest(type: String, trustCredential: String): JSONObject {
        return JSONObject()
            .put("type", type)
            .put("requestId", "req-${UUID.randomUUID()}")
            .put("deviceId", deviceId())
            .put("stableDeviceKey", stableDeviceKey())
            .put("displayName", Build.MODEL ?: "Android")
            .put("platform", "android")
            .put("architecture", System.getProperty("os.arch") ?: "unknown")
            .put("trustCredential", trustCredential)
    }

    private fun ensureTransport(
        host: String,
        port: Int,
        expectedFingerprint: String?,
        timeoutMs: Int,
    ): MsQuicNativeTransport = synchronized(transportLock) {
        val key = "$host:$port"
        transport?.let { existing ->
            if (endpointKey == key) {
                val observed = runCatching { existing.serverFingerprint() }.getOrNull()
                if (expectedFingerprint.isNullOrBlank() || expectedFingerprint.equals(observed, ignoreCase = true)) {
                    return@synchronized existing
                }
                existing.close()
                transport = null
            } else {
                existing.close()
                transport = null
            }
        }
        stateMachine.disconnect()
        stateMachine.begin(key)
        try {
            MsQuicNativeTransport(
                host = host,
                port = port,
                expectedFingerprint = expectedFingerprint,
                eventHandler = { bytes -> eventExecutor.execute { handleEnvelope(bytes) } },
                disconnectedHandler = { reason -> handleDisconnected(reason) },
                timeoutMilliseconds = timeoutMs,
            ).also {
                transport = it
                endpointKey = key
                stateMachine.authenticated(key)
            }
        } catch (error: Exception) {
            stateMachine.fail(key, "GatewayConnectFailed", error.message ?: "QUIC 连接失败")
            throw error
        }
    }

    private fun handleEnvelope(bytes: ByteArray) {
        val desktop = activeDesktop ?: return
        try {
            val envelope = JSONObject(String(bytes, Charsets.UTF_8))
            if (envelope.optString("messageType") != GatewayMessageType.EVENT.wireValue) {
                throw IllegalStateException("GatewayEventEnvelopeInvalid")
            }
            val gatewayResponse = envelope.getJSONObject("payload")
            if (!gatewayResponse.optBoolean("ok")) {
                throw IllegalStateException(gatewayResponse.optString("message", "桌面端事件失败"))
            }
            val payload = gatewayResponse.getJSONObject("payload")
            when (payload.optString("eventType")) {
                "scheme.updated" -> handleSchemeUpdate(desktop, payload)
                "jsapi.request" -> handleJsApiRequest(desktop, payload)
                else -> logs.append("Warning", "GatewayEvent", "收到未知事件：${payload.optString("eventType")}")
            }
        } catch (error: Exception) {
            logs.append("Error", "GatewayEvent", error.message ?: "服务端事件处理失败")
        }
    }

    private fun handleSchemeUpdate(desktop: JSONObject, payload: JSONObject) {
        val descriptor = payload.optJSONObject("scheme") ?: return
        val eventId = payload.optString("eventId")
        if (eventId.isBlank()) return
        if (onSchemeEvent(desktop, descriptor, eventId)) {
            request(
                desktop.getString("host"),
                desktop.optInt("port", 48320),
                authorizedRequest("scheme-ack", desktop.getString("trustCredential")).put("eventId", eventId),
                expectedFingerprint = desktop.optString("gatewayFingerprint"),
            )
        }
    }

    private fun handleJsApiRequest(desktop: JSONObject, payload: JSONObject) {
        val requestId = payload.optString("requestId")
        val capability = payload.optString("capability")
        if (requestId.isBlank() || capability.isBlank()) return
        val source = payload.optJSONObject("source") ?: JSONObject()
        val sourceKey = when (source.optString("kind")) {
            "component" -> "component:${source.optString("componentId", "unknown")}"
            "plugin" -> "plugin:${source.optString("pluginId", "unknown")}"
            "system" -> "system"
            else -> "unknown"
        }
        val result = onJsApiEvent(capability, payload.optJSONObject("payload") ?: JSONObject(), requestId, sourceKey)
        request(
            desktop.getString("host"),
            desktop.optInt("port", 48320),
            authorizedRequest("jsapi-response", desktop.getString("trustCredential"))
                .put("requestId", requestId)
                .put("responseOk", result.optBoolean("ok"))
                .put("errorCode", result.opt("errorCode") ?: JSONObject.NULL)
                .put("message", result.opt("message") ?: JSONObject.NULL)
                .put("payload", result.opt("payload") ?: JSONObject.NULL),
            expectedFingerprint = desktop.optString("gatewayFingerprint"),
        )
    }

    private fun sendHeartbeat() {
        val desktop = activeDesktop ?: return
        if (!running.get()) return
        runCatching {
            request(
                desktop.getString("host"),
                desktop.optInt("port", 48320),
                authorizedRequest("heartbeat", desktop.getString("trustCredential")),
                timeoutMs = 5_000,
                expectedFingerprint = desktop.optString("gatewayFingerprint"),
            )
        }.onFailure { error ->
            logs.append("Warning", "GatewayHeartbeat", error.message ?: "桌面端心跳失败")
        }
    }

    private fun handleDisconnected(reason: String) {
        if (!running.get()) return
        logs.append("Warning", "GatewayConnection", reason)
        synchronized(transportLock) {
            transport = null
            endpointKey = null
        }
        stateMachine.disconnect()
        val desktop = activeDesktop ?: return
        scheduleReconnect(desktop)
    }

    private fun scheduleReconnect(desktop: JSONObject) {
        synchronized(transportLock) {
            if (!running.get() || reconnectFuture?.isDone == false) return
            val attempt = reconnectAttempt++
            val delay = ReconnectBackoff.delayMilliseconds(attempt)
            reconnectFuture = scheduler.schedule(
                {
                    reconnectFuture = null
                    if (!running.get()) return@schedule
                    runCatching { startSubscription(desktop) }
                        .onFailure { error ->
                            logs.append("Error", "GatewayReconnect", "第 ${attempt + 1} 次重连失败：${error.message ?: "桌面端不可达"}")
                            scheduleReconnect(desktop)
                        }
                },
                delay,
                TimeUnit.MILLISECONDS,
            )
        }
    }

    override fun close() {
        stopSubscription()
        eventExecutor.shutdownNow()
        scheduler.shutdownNow()
    }
}

File: apps/mobile/ios/Transport/MobileGatewayClient.swift
import Foundation
import UIKit

final class MobileGatewayClient {
    private let deviceId: () -> String
    private let logs: MobileLogStore
    private let eventQueue = DispatchQueue(label: "cc.onedesk.mobile.gateway-events")
    private let state = MobileConnectionStateMachine()
    private let lock = NSLock()
    private var transport: MsQuicTransport?
    private var endpointKey: String?
    private var activeDesktop: (KnownDesktop, String)?
    private var heartbeat: DispatchSourceTimer?
    private var reconnectAttempt = 0
    private var running = false

    var onSchemeEvent: ((KnownDesktop, String, JSONObject, String) -> Bool)?
    var onJsApiEvent: ((String, JSONObject, String, String) -> JSONObject)?

    init(deviceId: @escaping () -> String, logs: MobileLogStore) {
        self.deviceId = deviceId
        self.logs = logs
    }

    func request(
        host: String,
        port: Int,
        payload: JSONObject,
        timeoutMilliseconds: Int = 7_000,
        expectedFingerprint: String? = nil
    ) throws -> JSONObject {
        var request = payload
        let requestId = request.string("requestId").isEmpty
            ? "req-\(UUID().uuidString.lowercased())"
            : request.string("requestId")
        request["requestId"] = requestId
        let client = try ensureTransport(
            host: host,
            port: port,
            expectedFingerprint: expectedFingerprint,
            timeoutMilliseconds: timeoutMilliseconds)
        let envelope: JSONObject = [
            "protocolVersion": OneDeskProtocol.version,
            "messageType": "request",
            "messageId": requestId,
            "correlationId": NSNull(),
            "payload": request,
        ]
        let response = try JSONSupport.object(from: client.request(
            try JSONSupport.data(envelope),
            timeoutMilliseconds: timeoutMilliseconds))
        guard response.string("messageType") == "response",
              response.string("correlationId") == requestId,
              let result = response.object("payload") else {
            throw OneDeskMobileError.transport("GatewayResponseCorrelationMismatch")
        }
        return result
    }

    func authorizedRequest(type: String, credential: String) -> JSONObject {
        [
            "type": type,
            "requestId": "req-\(UUID().uuidString.lowercased())",
            "deviceId": deviceId(),
            "displayName": UIDevice.current.name,
            "platform": "ios",
            "architecture": Self.architecture,
            "trustCredential": credential,
        ]
    }

    func serverFingerprint() throws -> String {
        guard let value = lock.withLock({ transport?.observedFingerprint }), !value.isEmpty else {
            throw OneDeskMobileError.transport("GatewayCertificateMissing")
        }
        return value
    }

    func uploadLog(desktop: KnownDesktop, credential: String, entry: JSONObject) -> Bool {
        do {
            var payload = authorizedRequest(type: "logs", credential: credential)
            payload["logs"] = [entry]
            return try request(
                host: desktop.host,
                port: desktop.port,
                payload: payload,
                timeoutMilliseconds: 2_500,
                expectedFingerprint: desktop.gatewayFingerprint).bool("ok")
        } catch {
            return false
        }
    }

    func startSubscription(desktop: KnownDesktop, credential: String) throws {
        running = true
        activeDesktop = (desktop, credential)
        var payload = authorizedRequest(type: "subscribe", credential: credential)
        payload["logs"] = []
        let response = try request(
            host: desktop.host,
            port: desktop.port,
            payload: payload,
            expectedFingerprint: desktop.gatewayFingerprint)
        guard response.bool("ok") else {
            running = false
            throw OneDeskMobileError.transport(response.string("message"))
        }
        reconnectAttempt = 0
        let endpoint = "\(desktop.host):\(desktop.port)"
        if state.state.phase == .synchronizing { try state.synchronized(endpoint: endpoint) }
        startHeartbeat()
    }

    func stop() {
        running = false
        heartbeat?.cancel()
        heartbeat = nil
        activeDesktop = nil
        reconnectAttempt = 0
        state.disconnect()
        lock.withLock {
            transport?.close()
            transport = nil
            endpointKey = nil
        }
    }

    private func ensureTransport(
        host: String,
        port: Int,
        expectedFingerprint: String?,
        timeoutMilliseconds: Int
    ) throws -> MsQuicTransport {
        try lock.withLock {
            let key = "\(host):\(port)"
            if let existing = transport, endpointKey == key {
                let observed = existing.observedFingerprint
                if expectedFingerprint?.isEmpty != false ||
                    expectedFingerprint?.caseInsensitiveCompare(observed ?? "") == .orderedSame {
                    return existing
                }
                existing.close()
                transport = nil
            } else if let existing = transport {
                existing.close()
                transport = nil
            }

            state.disconnect()
            try state.begin(endpoint: key)
            do {
                let client = try MsQuicTransport(
                    host: host,
                    port: port,
                    expectedFingerprint: expectedFingerprint,
                    timeoutMilliseconds: timeoutMilliseconds,
                    eventHandler: { [weak self] data in
                        self?.eventQueue.async { self?.handleEnvelope(data) }
                    },
                    disconnectedHandler: { [weak self] reason in
                        self?.handleDisconnected(reason)
                    })
                transport = client
                endpointKey = key
                try state.authenticated(endpoint: key)
                return client
            } catch {
                state.fail(endpoint: key, code: "GatewayConnectFailed", message: error.localizedDescription)
                throw error
            }
        }
    }

    private func handleEnvelope(_ data: Data) {
        guard let (desktop, credential) = activeDesktop else { return }
        do {
            let envelope = try JSONSupport.object(from: data)
            guard envelope.string("messageType") == "event",
                  let gatewayResponse = envelope.object("payload"),
                  gatewayResponse.bool("ok"),
                  let payload = gatewayResponse.object("payload") else {
                throw OneDeskMobileError.invalidPayload("GatewayEventEnvelopeInvalid")
            }
            switch payload.string("eventType") {
            case "scheme.updated":
                guard let descriptor = payload.object("scheme") else { return }
                let eventId = payload.string("eventId")
                if onSchemeEvent?(desktop, credential, descriptor, eventId) == true {
                    var ack = authorizedRequest(type: "scheme-ack", credential: credential)
                    ack["eventId"] = eventId
                    _ = try request(host: desktop.host, port: desktop.port, payload: ack,
                                    expectedFingerprint: desktop.gatewayFingerprint)
                }
            case "jsapi.request":
                try handleJsApiRequest(desktop: desktop, credential: credential, payload: payload)
            default:
                logs.append("Warning", "GatewayEvent", "收到未知事件：\(payload.string("eventType"))")
            }
        } catch {
            logs.append("Error", "GatewayEvent", error.localizedDescription)
        }
    }

    private func handleJsApiRequest(desktop: KnownDesktop, credential: String, payload: JSONObject) throws {
        let requestId = payload.string("requestId")
        let capability = payload.string("capability")
        guard !requestId.isEmpty, !capability.isEmpty else { return }
        let source = payload.object("source") ?? [:]
        let sourceKey: String
        switch source.string("kind") {
        case "component": sourceKey = "component:\(source.string("componentId"))"
        case "plugin": sourceKey = "plugin:\(source.string("pluginId"))"
        case "system": sourceKey = "system"
        default: sourceKey = "unknown"
        }
        let result = onJsApiEvent?(capability, payload.object("payload") ?? [:], requestId, sourceKey)
            ?? ["ok": false, "errorCode": "CapabilityNotSupported", "message": "iOS 能力未实现"]
        var response = authorizedRequest(type: "jsapi-response", credential: credential)
        response["requestId"] = requestId
        response["responseOk"] = result.bool("ok")
        response["errorCode"] = result["errorCode"] ?? NSNull()
        response["message"] = result["message"] ?? NSNull()
        response["payload"] = result["payload"] ?? NSNull()
        _ = try request(host: desktop.host, port: desktop.port, payload: response,
                        expectedFingerprint: desktop.gatewayFingerprint)
    }

    private func startHeartbeat() {
        heartbeat?.cancel()
        let timer = DispatchSource.makeTimerSource(queue: eventQueue)
        timer.schedule(deadline: .now() + 15, repeating: 15)
        timer.setEventHandler { [weak self] in self?.sendHeartbeat() }
        heartbeat = timer
        timer.resume()
    }

    private func sendHeartbeat() {
        guard running, let (desktop, credential) = activeDesktop else { return }
        do {
            let payload = authorizedRequest(type: "heartbeat", credential: credential)
            _ = try request(host: desktop.host, port: desktop.port, payload: payload,
                            timeoutMilliseconds: 5_000, expectedFingerprint: desktop.gatewayFingerprint)
        } catch {
            logs.append("Warning", "GatewayHeartbeat", error.localizedDescription)
        }
    }

    private func handleDisconnected(_ reason: String) {
        guard running else { return }
        logs.append("Warning", "GatewayConnection", reason)
        lock.withLock {
            transport = nil
            endpointKey = nil
        }
        scheduleReconnect()
    }

    private func scheduleReconnect() {
        guard running, let (desktop, credential) = activeDesktop else { return }
        let delay = min(30.0, pow(2.0, Double(reconnectAttempt)))
        reconnectAttempt += 1
        eventQueue.asyncAfter(deadline: .now() + delay) { [weak self] in
            guard let self, self.running else { return }
            do {
                try self.startSubscription(desktop: desktop, credential: credential)
            } catch {
                self.logs.append("Error", "GatewayReconnect", error.localizedDescription)
                self.scheduleReconnect()
            }
        }
    }

    private static var architecture: String {
        #if arch(arm64)
        return "arm64"
        #elseif arch(x86_64)
        return "x86_64"
        #else
        return "unknown"
        #endif
    }
}
